File: source/material/Material.java
package material;

import callback.EmptyCallback;
import glparts.Shaders;
import glparts.Texture;
import math.Float3;
import named.NameHolder;
import named.Named;
import utility.Instance;

import java.io.Serializable;

public class Material extends Named implements Serializable {
    public final Float3 color = new Float3(1);
    public float roughness = 0.5f;

    public Texture colorMap = null;
    public Texture normalMap = null;
    public Texture roughnessMap = null;

    public Material(NameHolder holder, String name) {
        super(holder, name);
    }

    public void use(EmptyCallback callback) {
        int[] textures = new int[3];
        textures[0] = Instance.isValid(colorMap) ? colorMap.getBuffer() : 0;
        textures[1] = Instance.isValid(normalMap) ? normalMap.getBuffer() : 0;
        textures[2] = Instance.isValid(roughnessMap) ? roughnessMap.getBuffer() : 0;
        Texture.use(textures, callback);
    }

    public void updateUniforms(Shaders shaders) {
        final boolean useColorMap = Instance.isValid(colorMap);
        shaders.setUniform("color", color);
        shaders.setUniform("useColorMap", useColorMap ? 1 : 0);
        shaders.setUniform("colorMap", 0);
    }
}

File: source/entity/material/Material.kt
package entity.material

import glparts.Shaders
import glparts.Texture
import math.Float3
import named.NameHolder
import named.Named
import java.io.Serializable

class Material(holder: NameHolder, name: String) : Named(holder, name), Serializable {
    var colorBlend: Float = 0f
    var roughness: Float = 0.5f

    var color: Float3 = Float3(1f)

    var colorMap: Texture? = null
    var roughnessMap: Texture? = null
    var normalMap: Texture? = null

    fun use(callback: () -> Unit) {
        val textures = IntArray(3)
        textures[0] = colorMap?.buffer ?: 0
        textures[1] = normalMap?.buffer ?: 0
        textures[2] = roughnessMap?.buffer ?: 0
        Texture.use(textures, callback)
    }

    fun updateUniforms(shaders: Shaders) {
        shaders.setUniform("color", color)
        shaders.setUniform("colorMap", 0)
        shaders.setUniform("colorBlend", colorBlend)
    }
}